akka-remote/src/main/scala/akka/remote/artery/Handshake.scala

```diff
@@ -162,12 +125,33 @@ private[akka] class InboundHandshake(inboundContext: InboundContext) extends Gra
       // InHandler
       override def onPush(): Unit = {
         grab(in) match {
-          case InboundEnvelope(_, _, HandshakeReq(from), _) ⇒
-            inboundContext.association(from.address).completeRemoteAddress(from)
+          case InboundEnvelope(_, _, HandshakeReq(from), _, _) ⇒
+            inboundContext.association(from.address).completeHandshake(from)
             inboundContext.sendControl(from.address, HandshakeRsp(inboundContext.localAddress))
             pull(in)
+          case InboundEnvelope(_, _, HandshakeRsp(from), _, _) ⇒
+            inboundContext.association(from.address).completeHandshake(from)
+            pull(in)
           case other ⇒
-            push(out, other)
+            if (isKnownOrigin(other.originAddress))
```

Comment:
@drewhk When implementing this part (detecting message from unknown origin) I realized that the idea with only including the the UID in each message instead of the full originAddress will cause some trouble. When detecting that we need to initiate a new handshake by sending back a control message to the origin address, but with the UID only we don't have the address.

I see a few alternatives:
1) Always include full origin address
2) Send `HandshakeReq` once in while (e.g. once per second) for active associations. Drop messages from unknown UID but when the `HandshakeReq` arrives the new handshake exchange kicks in.